internal/dag/gatewayapi_processor_test.go

```diff
@@ -27,166 +27,251 @@ import (
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & wildcard hostname and matching hostname match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+				"http.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & wildcard hostname and non-matching hostname don't match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+				"not.matching.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: []error{fmt.Errorf("gateway hostname \"http.projectcontour.io\" does not match hostnames hostname \"not.matching.io\"")},
+		},
+		"listener host wildcard & wildcard hostnames host match": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"*.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener wildcard matchall host & host match": {
+			listenerHost: listenerHostname("*"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener wildcard matchall host & multiple host match": {
+			listenerHost: listenerHostname("*"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+				"http2.projectcontour.io",
+				"http3.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io":  {},
+				"http2.projectcontour.io": {},
+				"http3.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & hostname not defined match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames:    []gatewayapi_v1alpha1.Hostname{},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host with many labels doesn't match hostnames wildcard host": {
+			listenerHost: listenerHostname("too.many.labels.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want:      map[string]struct{}{},
+			wantError: []error{fmt.Errorf("hostnames hostname \"*.projectcontour.io\" does not match gateway hostname \"too.many.labels.projectcontour.io\"")},
+		},
+		"listener wildcard host doesn't match hostnames with many labels host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"too.many.labels.projectcontour.io",
+			},
+			want:      map[string]struct{}{},
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match hostnames hostname \"too.many.labels.projectcontour.io\"")},
```

Comment:
maybe add a test or two where a route tries to define a hostname `*` with the listener also allowing `*`

Reply:
ok added a test with `*` on listener with `*.projectcontour.io` on route.

Reply:
sorry to be a pain but I think a test where the HTTPRoute tries to use `*` as a hostname should be added too, since that is invalid

I don't remember if somewhere else in the code already restricts us from using that but if so would be good to have an explicit test somewhere, here or in the builder or something?